remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java

```diff
@@ -0,0 +1,659 @@
+            deleteKeys(logFiles);
+            log.debug("Deleting data files: {}", dataFiles);
+            deleteKeys(dataFiles);
+            log.debug("Deleting last modified reverse index entries: {}", lastModifiedReverseIndexEntries);
+            deleteKeys(lastModifiedReverseIndexEntries);
+        } catch (SdkClientException e) {
+            throw new KafkaException("Error cleaning log until " + cleanUpTillMs + " in " + topicPartition, e);
+        }
+
+        // for now, return the earliest offset after deletion.
+        // TODO confirm it's the right way
+        return earliestLogOffset(topicPartition);
+    }
+
+    private ListObjectsV2Request createListObjectsRequest(String prefix) {
+        return new ListObjectsV2Request()
+                .withBucketName(bucket)
+                .withPrefix(prefix)
+                .withMaxKeys(maxKeys); // it's ok to pass null here
+    }
+
+    private void deleteKeys(Collection<String> keys) {
+        log.debug("Deleting keys {}", keys);
+        List<List<String>> chunks = new ArrayList<>();
+        chunks.add(new ArrayList<>());
+
+        for (String key : keys) {
+            List<String> lastChunk = chunks.get(chunks.size() - 1);
+            if (lastChunk.size() >= 1000) {
+                lastChunk = new ArrayList<>();
+                chunks.add(lastChunk);
+            }
+            lastChunk.add(key);
+        }
+
+        for (List<String> chunk : chunks) {
+            assert chunk.size() <= 1000;
+
+            final String[] chunkArray = chunk.toArray(new String[]{});
+            log.debug("Deleting key chunk {}", Arrays.toString(chunkArray));
+            DeleteObjectsRequest deleteObjectsRequest = new DeleteObjectsRequest(bucket)
+                    .withKeys(chunkArray);
+            try {
+                s3Client.deleteObjects(deleteObjectsRequest);
+            } catch (MultiObjectDeleteException e) {
+                // On an attempt to delete a non-existent key, real S3 will return no error,
+                // but Localstack (used for testing) will. This effectively handles errors that
+                // appear only in tests with Localstack.
+                // TODO not needed if we switch to integration testing with real S3.
+                for (MultiObjectDeleteException.DeleteError error : e.getErrors()) {
+                    log.warn("Error deleting key {}: {} {}",
+                            error.getKey(), error.getCode(), error.getMessage());
+                }
+            } catch (SdkClientException e) {
+                throw new KafkaException("Error deleting keys " + Arrays.toString(chunkArray), e);
+            }
+        }
+    }
+
+    @Override
+    public Records read(RemoteLogIndexEntry remoteLogIndexEntry,
+                        int maxBytes,
+                        long startOffset,
+                        boolean minOneMessage) throws IOException {
+        // TODO what to return in case nothing exists for this offset?
```

Comment:
Good point, I guess we should throw  `OffsetOutOfRangeException`. WDYT?

Reply:
@satishd `OffsetOutOfRangeException` seems to be a client exception. We can create a new exception type for this or stick to returning `MemoryRecords.EMPTY`.